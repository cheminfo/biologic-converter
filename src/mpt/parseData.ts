import { MeasurementVariable } from 'cheminfo-types';

import { getOneLetter } from '../utility/getOneLetter';
import { dataColumnsByName } from '../utility/ids';

interface Data {
  [variableName: string]: MeasurementVariable;
}
/**
 * Parses the data from the MPT file
 * @param data - string[] sliced where data starts
 * @returns - the data as an object, keys are the names of the data-fields
 */
export function parseData(data: string[]): Data {
  const variables: Data = {};

  let matrix = data.map((line) => line.trim().split('\t'));

  const fields = matrix[0];

  matrix = matrix.slice(1);
  for (let i = 0; i < fields.length; i++) {
    const fieldName = fields[i];
    if (fieldName === '') continue;
    const oneLetter = getOneLetter(i);
    const { name, unit } = mptNameToMPRName(fieldName);
    variables[oneLetter] = {
      label: name,
      units: unit,
      isDependent: fieldName !== 'time/s',
      data: matrix.map((row) => Number(row[i])),
    };
  }
  return variables;
}

function mptNameToMPRName(fieldName: string) {
  const mapToLabelUnit = dataColumnsByName[fieldName];
  if (mapToLabelUnit !== undefined) {
    return mapToLabelUnit;
  } else {
    const [name, unit] = fieldName.split('/') || [fieldName, ''];
    return { name, unit };
  }
}
