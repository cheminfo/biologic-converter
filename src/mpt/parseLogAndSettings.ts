import { OutParams, getParams } from '../utility/getParamsFromText';
import { normalizeFlag, normalizeKeyValue } from '../utility/normalize';
import type { Technique } from '../utility/techniqueFromId';

import { addKVToObject } from './utility/addKeyWithoutOverwrite';

export interface LogAndSettings {
  settings: {
    variables: {
      //similar to the ParseSettings of MPR format
      technique: string;
      params: OutParams;
      flags: string[];
      modified: { on: string; newParams: OutParams }[];
      /* looking for a better definition */
      [key: string]: any;
    };
  };
  log: {
    variables: {
      flags: string[];
      /* looking for a better definition */
      [key: string]: any;
    };
  };
}

/**
 * Parses log and settings modules of text files
 * @param lines - file as string[]
 * @param technique - the technique being processed,
 * used to format the parameters to MPR format
 * @returns JSON object representing the parsed data
 */
export function parseLogAndSettings(
  lines: string[],
  technique: Technique,
): LogAndSettings {
  const result: LogAndSettings = makeBaseObject(technique.name);

  const regex = {
    isEmpty: /^\s*$/,
    isKeyValue: / : | :$/,
    isMultiline: /^[\t ]/,
    isParameters: /\s{3,}$/,
  };

  for (let i = 0; i < lines.length; i++) {
    const currentLine = lines[i];
    if (regex.isEmpty.test(currentLine)) {
      continue;
    } else if (regex.isKeyValue.test(currentLine)) {
      // eslint-disable-next-line prefer-named-capture-group
      let [key, val] = currentLine.split(/ :(.*)/);
      key = key.trim();
      val = val.trim();
      if (regex.isMultiline.test(lines[i + 1])) {
        do {
          //just a few short text lines
          val.concat('\n', lines[++i]);
        } while (regex.isMultiline.test(lines[i + 1]));
      }
      const [newKey, logOrSettings, newVal] = normalizeKeyValue(key, val);
      addKVToObject(result[logOrSettings].variables, newKey, newVal);
    } else if (regex.isParameters.test(currentLine)) {
      const currentVariables = result.settings.variables;
      if (Object.keys(currentVariables.params).length === 0) {
        const [params, lastLineRead] = getParams(
          technique.preParameters,
          lines,
          i,
        );
        currentVariables.params = params;
        i = lastLineRead;
      } else {
        const modified = lines[i - 1].match(/Modify on : (?<date>.*)/);
        if (modified?.groups?.date) {
          const [newParams, lastLineRead] = getParams(
            technique.preParameters,
            lines,
            i,
          );
          currentVariables.modified.push({
            on: modified.groups.date,
            newParams,
          });
          i = lastLineRead;
        }
      }
    } else {
      const [theKey, logOrSettings, theValue] = normalizeFlag(
        currentLine.trim(),
      );
      if (theValue) {
        result[logOrSettings].variables[theKey] = theValue;
      } else {
        result[logOrSettings].variables.flags.push(theKey);
      }
    }
  }
  return result;
}

function makeBaseObject(techniqueName: string) {
  return {
    settings: {
      variables: {
        technique: techniqueName,
        params: {},
        flags: [],
        modified: [],
      },
    },
    log: {
      variables: {
        flags: [],
      },
    },
  };
}
