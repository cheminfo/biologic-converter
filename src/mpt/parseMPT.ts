import { MeasurementVariable, TextData } from 'cheminfo-types';
import { ensureString } from 'ensure-string';

import { ComplexObject } from '../Types';
import { parseMeta } from '../parseMeta';

export interface MPT {
  /* settings */
  meta: ComplexObject;
  /* body, i.e results */
  variables: Record<string, MeasurementVariable>;
}

/**
 * Parses BioLogic MPT files
 * @param data - as a string, Buffer, ArrayBuffer.
 * @returns JSON Object with parsed data
 */
export function parseMPT(data: TextData): MPT {
  const lines = ensureString(data, {
    encoding: 'latin1',
  }).split(/\r?\n/);

  const header = headerMPT(lines.slice(0, 4));
  const i = parseInt(header['Nb header lines'], 10) - 2;
  return {
    meta: Object.assign(header, parseMeta(lines.slice(4, i))),
    variables: parseData(lines.slice(i + 1)),
  };
}

/**
 * Parse the values
 */
function parseData(data: string[]): MPT['variables'] {
  const variables: Record<string, MeasurementVariable> = {};

  let matrix = data.map((line) => line.split('\t'));

  const fields = matrix[0];

  matrix = matrix.slice(1);
  for (let i = 0; i < fields.length; i++) {
    const fieldName = fields[i];
    if (!fieldName.includes('/')) continue;
    variables[fieldName] = {
      label: fieldName.split('/')[0],
      units: fieldName.split('/')[1],
      isDependent: fieldName !== 'time/s',
      data: matrix.map((row) => Number(row[i])),
    };
  }
  return variables;
}

interface HeaderMPT {
  fileType: string;
  Technique: string;
  [other: string]: string;
}

function headerMPT(lines: string[]): HeaderMPT {
  const nOfLinesHeader = lines[1].split(' : ');
  return {
    fileType: lines[0],
    'Nb header lines': nOfLinesHeader[1].trim() || '',
    Technique: lines[3],
  };
}
