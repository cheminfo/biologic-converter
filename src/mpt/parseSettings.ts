import { ComplexObject } from '../Types';
import { normalizeKeyValue } from '../utility/normalize';
/**
 * Creates an mps object from an mps file
 * The output is similar, but not the same, than `MPR.settings`
 * MPS includes one or more techniques
 * We will know what this means when using it.
 *
 * @param data - pass the file as string, Buffer or Arraybuffer.
 * @returns JSON object representing the parsed data
 */

export function parseSettings(
  lines: string[],
  technique: string,
): ComplexObject {
  // file converted to an array of strings, each item a newline.

  let result: ComplexObject = { technique, params: [] };

  const regex = {
    nothing: /^\s*$/,
    keyValue: / : | :$/,
    multiline: /^[ \t]/,
    table: /^\w.*\s{2,}-*\w+.*\s{4,}$/,
  };

  for (let i = 0; i < lines.length; i++) {
    const currentLine = lines[i];

    if (regex.nothing.test(currentLine)) {
      continue;
    } else if (regex.keyValue.test(currentLine)) {
      const kV: string[] = currentLine.split(regex.keyValue); //if many " : " we fix below
      //function updates the index bc some values are multiline
      let key: string = kV[0].trim();
      let val: string = kV.slice(1).join(' : ').trim();
      while (regex.multiline.test(lines[i + 1])) {
        //use original value then
        val = val.concat('\n', lines[++i].trim());
      }
      const [newKey, newVal] = normalizeKeyValue(key, val);
      result[newKey] = newVal;
    } else if (regex.table.test(currentLine)) {
      /* for not k : v */
      //regex.table
      const kV: string[] = currentLine.split(/\s{2,}/);
      const [key, val] = [kV[0].trim(), kV.slice(1).join('  ').trim()];
      result[key] = val;
    } else if (Array.isArray(result.flags)) {
      result.flags.push(currentLine);
    } else {
      result.flags = [currentLine];
    }
  }
  return result;
}
