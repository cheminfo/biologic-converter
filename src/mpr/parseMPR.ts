import { BinaryData } from 'cheminfo-types';
import { IOBuffer } from 'iobuffer';

import { parseData, ParseData } from './modules/parseData';
import { parseLogs, ParseLogs } from './modules/parseLogs';
import { parseLoop, ParseLoop } from './modules/parseLoop';
import { parseModuleHeader, ModuleHeader } from './modules/parseModuleHeader';
import { parseSettings, ParseSettings } from './modules/parseSettings';
import { isModule } from './utility/isModule';

export interface MPR {
  name: string /** a string in the first line */;
  data: { header: ModuleHeader; variables: ParseData };
  settings: { header: ModuleHeader; variables: ParseSettings };
  log?: { header: ModuleHeader; variables: ParseLogs };
  loop?: { header: ModuleHeader; variables: ParseLoop };
}

/**
 * Parses binary `.mpr` files
 * arrayBuffer - the data itself.
 * @returns the data as a JSON-like object.
 */
export function parseMPR(arrayBuffer: BinaryData): MPR {
  const buffer = new IOBuffer(arrayBuffer);
  let mpr: Partial<MPR> = {};

  // top level properties
  mpr.name = buffer
    .readUtf8(0x34)
    .replace(/\x1A|\x00/g, '')
    .trim();

  while (isModule(buffer)) {
    const header = parseModuleHeader(buffer); //this is added to the objects below
    const dataStart = buffer.offset;
    const dataLength = header.length;

    if (/settings/i.test(header.longName)) {
      mpr.settings = { header, variables: parseSettings(buffer) };
    } else if (/data/i.test(header.longName)) {
      mpr.data = { header, variables: parseData(buffer, header) };
    } else if (/log/i.test(header.longName)) {
      mpr.log = { header, variables: parseLogs(buffer) };
    } else if (/loop/i.test(header.longName)) {
      mpr.loop = { header, variables: parseLoop(buffer) };
    }
    buffer.offset = dataStart + dataLength; //end of module
  }
  return mpr as MPR;
}
