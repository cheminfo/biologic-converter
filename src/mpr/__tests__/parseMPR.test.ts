import { readFileSync } from 'node:fs';
import { join } from 'node:path';

import { MPT, parseMPT } from '../../mpt/parseMPT';
import { parseMPR } from '../parseMPR';

const testFiles = join(__dirname, '/../../__tests__/data/all/zir');

describe('parse zir data', () => {
  const mprBuffer = readFileSync(join(testFiles, 'zir.mpr'));
  const mptBuffer = readFileSync(join(testFiles, 'zir.mpt'));

  const mpr = parseMPR(mprBuffer);
  const variables = mpr.data.variables;

  const mpt = parseMPT(mptBuffer) as Required<MPT>;
  const mptVariables = mpt.data?.variables;
  it('time', () => {
    const mprTime = variables.f;
    const mptTime = mptVariables.f;
    expect(mprTime.data).toHaveLength(mptTime.data.length);
    expect(mptTime).toMatchObject(mprTime);
    expect(mprTime.data[0]).toBeCloseTo(mptTime.data[0], 5);
  });
  it('freq', () => {
    const mprFreq = variables.a;
    const mptFreq = mptVariables.a;
    expect(mprFreq.data).toHaveLength(mptFreq.data.length);
    expect(mprFreq.data[0]).toBeCloseTo(mptFreq.data[0], 3);
  });
  it('Re(Z)', () => {
    const mprRe = variables.b;
    const mptRe = mptVariables.b;
    expect(mprRe.data).toHaveLength(mptRe.data.length);
    expect(mprRe.data[0]).toBeCloseTo(mptRe.data[0], 2);
  });
  it('I Range', () => {
    const mprRe = variables.g;
    const mptRe = mptVariables.g;
    expect(mprRe.data).toHaveLength(mptRe.data.length);
    expect(mprRe.data[0]).toBeCloseTo(mptRe.data[0], 2);
  });
});
