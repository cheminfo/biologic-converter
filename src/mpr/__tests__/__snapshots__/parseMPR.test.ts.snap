// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compare the technique (settings only) cp meta 1`] = `
{
  "acquisitionStart": 0,
  "activeMaterialMass": 0.0010000000474974513,
  "atX": 0,
  "atomicWeight": 0.0010000000474974513,
  "batteryCapacity": 3.615350037958028e-43,
  "batteryCapacityUnit": 0,
  "characteristicMass": 0.0010000000474974513,
  "comments": "",
  "eTransferred": 1,
  "electrodeMaterial": "",
  "electrodeSurfaceArea": 0.0010000000474974513,
  "electrolyte": "",
  "molecularWeight": 0.0010000000474974513,
  "params": {
    "Bandwidth": 8,
    "E range max": 10,
    "E range min": -10,
    "EM": "pass",
    "I Range": "1 A",
    "Is": -100,
    "Is vs.": 2,
    "dEs": 0,
    "dQM": 3.3329999446868896,
    "dts": 1,
    "goto Ns": 0,
    "nc cycles": 0,
    "record": 1,
    "ts": 120,
    "unit Is": "mA",
    "unit dQM": 1,
  },
  "referenceElectrode": "SCE Saturated Calomel Electrode",
  "technique": "CP",
}
`;
